Fix flaky subscription and spinning test

wjwwood opened the pull request:
This pr does two things, first it adds a way to print out a meaningful string for a given FutureReturnCode (which I needed when debugging). You can now get an output like this:

```
/Users/william/ros2_ws/src/ros2/system_tests/test_rclcpp/test/test_subscription.cpp:80: Failure
Value of: future_ret
  Actual: INTERRUPTED (1)
Expected: FutureReturnCode::SUCCESS
Which is: SUCCESS (0)
```

Second it refactors the `executor.spin_until_future_complete` which was, imo, not working correctly before and which I needed when addressing some flaky-ness of one of the tests in `test_rclcpp`. Before this patch, I believe it would pass the full `timeout` value to `spin_once` even if it had spent some time spinning a previous time through the while loop. The changes to fix up the test (and use these changes) is forthcoming.

Connects to ros2/system_tests#108

wjwwood commented:
This PR is attached to https://github.com/ros2/system_tests/pull/108, can I get some reviews of this?

dirk-thomas commented on rclcpp/include/rclcpp/executor.hpp:
```diff
@@ -177,23 +184,30 @@ class Executor
+      }
+      // Otherwise check if we still have time to wait, return TIMEOUT if not.
+      auto elapsed_time = std::chrono::steady_clock::now() - start_time;
+      if (elapsed_time > timeout) {
```
Should we check for `>=` here?

jacquelinekay replied:
For this comparison to work, don't you need to store the initial value of `elapsed_time`? Otherwise you could exit early.

wjwwood replied:
Yeah, it's just wrong... I'm gonna refactor it again and write a test.

dirk-thomas commented on rclcpp/include/rclcpp/executor.hpp:
```diff
@@ -177,23 +184,30 @@ class Executor
+        return FutureReturnCode::TIMEOUT;
+      }
+      // Subtract the elapsed time from the original timeout.
+      timeout -= std::chrono::duration_cast<std::chrono::duration<int64_t, TimeT>>(elapsed_time);
```
Why is `timeout` reduced here? Doesn't that make the loop compare wrongly on the next iteration of line 203?

wjwwood replied:
Yeah, it does. Let me refactor this real quick.

dirk-thomas commented on rclcpp/src/rclcpp/executor.cpp:
```diff
@@ -569,3 +572,25 @@ Executor::get_next_executable(std::chrono::nanoseconds timeout)
+    case FutureReturnCode::INTERRUPTED:
+      return string("INTERRUPTED (" + to_string(static_cast<enum_type>(future_return_code)) + ")");
+    case FutureReturnCode::TIMEOUT:
+      return string("TIMEOUT (" + to_string(static_cast<enum_type>(future_return_code)) + ")");
```
Needs `default`.

wjwwood replied:
Actually it doesn't since this is an Enum switch. The compiler will warn you if the enum is extended in the future that not all cases are covered.

wjwwood commented:
Ok, let's try this again!

I added a test for this particular function in which the last case tested failed with the code you guys tried to review before. I also refactored the function (again) and now it passes the tests. So this is ready for re-review.

dirk-thomas commented:
The algo lgtm now. +1
